### PR TITLE
Add Alpine Linux support

### PR DESCRIPTION
**Pull request by pablogsal**
Closes: #74

### CODE REVIEW

File: src/memray/_memray/tracking_api.h [hunk 1]
```diff
@@ -23,14 +23,35 @@
 #include "record_writer.h"
 #include "records.h"
 
-#ifdef MEMRAY_TLS_MODEL
+// Only define use a different TLS model than the default if we are using glibc
+// as is the only implementation that can give us some space in the static TLS
+// array to place our TLS with confidence. In general, in other implementations
+// this may result in a relocation error at compilation time or at load time.
```
Comment (godlygeek) [inline src/memray/_memray/tracking_api.h:29]: I _think_ `local-dynamic` should be safe even on musl.

File: src/memray/_memray/hooks.cpp [hunk 2]
```diff
@@ -173,7 +175,11 @@ memalign(size_t alignment, size_t size) noexcept
 {
     assert(hooks::memalign);
 
-    void* ret = hooks::memalign(alignment, size);
+    void* ret;
+    {
+        tracking_api::RecursionGuard guard;
```
Comment (godlygeek) [inline src/memray/_memray/hooks.cpp:180]: Is this an unrelated change? Or is this necessary because `memalign` in musl calls into a function that we've intercepted?

File: src/memray/_memray/alloc.h [hunk 3]
```diff
@@ -0,0 +1,10 @@
+#include <malloc.h>
+#include <stdlib.h>
+
+#ifndef __GLIBC__
+void*
+pvalloc(size_t size)
```
Comment (godlygeek) [inline src/memray/_memray/alloc.h:6]: This would break and require a change to our source if musl ever added `pvalloc`. But fixing that requires some sort of configure-like system, or feature test macros, or something... 😕

File: src/memray/_memray/hooks.cpp (hunk 2 above)
Comment (pablogsal) [reply src/memray/_memray/hooks.cpp:180]: Is exactly because of that. `muslc` doesn't use private symbols in this case. I didn't want to add this everywhere to not impact performance.

File: src/memray/_memray/alloc.h (hunk 3 above)
Comment (pablogsal) [reply src/memray/_memray/alloc.h:6]: If you want to handle this differently, I am open to suggestions :)

Comment (pablogsal) [reply src/memray/_memray/alloc.h:6]: Technically we just need this so the test suite works, but maybe there are other better ways to do this...

Comment (godlygeek) [reply src/memray/_memray/alloc.h:6]: The test suite probably shouldn't depend on a function that even the glibc docs document as "obsolete", to be fair...

We're also not intercepting `aligned_alloc` right now, which the glibc docs suggest as a replacement for the obsolete `valloc`... I opened #78 for that.

File: src/memray/_memray/hooks.cpp (hunk 2 above)
Comment (godlygeek) [reply src/memray/_memray/hooks.cpp:180]: Hm - would it be better to just move the `RecursionGuard` handling up from `trackAllocationImpl` and `trackDeallocationImpl` into each of the hook functions? 

It would be more consistent, at the cost of a few lines of copy pasted code.

> We only do this (for now) in this call to avoid the extra overhead of the recursion guard in the rest of the allocators (that are more commonly used by far).

> I didn't want to add this everywhere to not impact performance.

What performance are you worried about? This is only hit if the hooks are installed, and in the happy path the call to `valloc` succeeds and we call into `trackAllocation` and it sets the `RecursionGuard` itself. If we hoist the creation of the `RecursionGuard` out of `trackAllocationImpl` and `trackDeallocationImpl` and into the hook functions, then it gets used the same number of times as it used to, so it shouldn't have any performance impact at all. We're already turning it on, we just want to turn it on a bit earlier than we have been.

Or am I missing something?

File: src/memray/_memray/alloc.h (hunk 3 above)
Comment (godlygeek) [reply src/memray/_memray/alloc.h:6]: Alright, I can't immediately think of any better idea than this, but - doesn't this need an `extern "C"` on it?

File: src/memray/_memray/elf_utils.h [hunk 4]
```diff
@@ -38,18 +38,32 @@ using Sym = ElfW(Sym);
...
+ensureRelocatedAddress(const Addr base, const U addr)
+{
+    // Depending on the libc version, some values may be already relocated
+    // or not. So we need to check first if the relocation already happened
+    // and make it ourselves if that is not the case.
+    auto the_addr = reinterpret_cast<Addr>(addr);
+    if (the_addr < base) {
+        return reinterpret_cast<T>(base + the_addr);
+    }
+    return reinterpret_cast<T>(the_addr);
```
[the remark this answers is not in the document]
Comment (pablogsal) [reply src/memray/_memray/elf_utils.h:52]: Nothing as far as I know, but it is extremely unlikely. Take into account that `base` is the address of the dynamic section of the elf object and just the load point for 64 bits is already 0x40000 (and that's the entry point, you need to consider anything after it and before the dynamic section table).

File: src/memray/_memray/alloc.h (hunk 3 above)
Comment (pablogsal) [reply src/memray/_memray/alloc.h:6]: We *could* just not care about `pvalloc` to be honest....

File: src/memray/_memray/hooks.cpp (hunk 2 above)
Comment (godlygeek) [reply src/memray/_memray/hooks.cpp:180]: OK. I think this is a bit ugly, and I'd like to see if we can come up with something more elegant, but that can be left for another PR.

File: src/memray/_memray/alloc.h (hunk 3 above)
Comment (godlygeek) [reply src/memray/_memray/alloc.h:6]: I think it's fine, let's just add the `extern "C"`

**Review by godlygeek — APPROVED** [review]
LGTM - The RecursionGuard handling is a bit ugly, but we can clean it up in a follow up PR. This definitely moves the ball forward.
